### ISSUE TITLE
Problems writing TIFF files

### ISSUE THREAD

**Issue by evillrich** [haraldk/TwelveMonkeys#226]
Hi,

I just started using TwelveMonkeys 3.2.1 earlier today to work with some TIFF images in my application. I really like the approach taken by the library - pure Java ImageIO plugins.  I have successfully read various TIFF images so I'm happy with that aspect of the library.

However, I can't seem to write good TIFF files out.  Once I read a TIFF in, I can write it out as a JPEG successfully.  Here is a simple app using some images available on the Internet:

``````java
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

import javax.imageio.ImageIO;

// Images:
// http://www.fileformat.info/format/tiff/sample/CCITT_1.TIF
// http://www.fileformat.info/format/tiff/sample/G31DS.TIF

public class WriteTiffTest {
    public static void main(String args[]) throws IOException {
        BufferedImage image = ImageIO.read(new File(args[0]));

        // Write that TIFF as a JPEG - works fine
        if (!ImageIO.write(image, "JPEG", new File("test.jpg"))) {
            System.out.println("Could not write out JPEG image");
        }

        // Now write that TIFF as another TIFF - image is empty
        if (!ImageIO.write(image, "TIFF", new File("test.tif"))) {
            System.out.println("Could not write out TIFF image");
        }

        image = ImageIO.read(new File("test.tif"));

        // Now write that TIFF as a JPEG
        if (!ImageIO.write(image, "JPEG", new File("test2.jpg"))) {
            System.out.println("Could not write out JPEG image");
        }
    }
}
``````

Note that the TIFF file is output but can't be viewed by many popular TIFF editors.  Also, if I perform some other image manipulation (e.g. cropping) and write the file out as a tiff and later read the file in, the image data appears corrupted.

These appear to be TIFF formats supported by TwelveMonkeys per the documentation.

I'm using JDK 7.  Note that I also tried using the 3.0.2 just to have another data point.  Unfortunately, that version doesn't support reading the TIFF images - I got the following exception:

``````
Exception in thread "main" javax.imageio.IIOException: Unsupported TIFF Compression value: 4
	at com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader.read(Unknown Source)
	at javax.imageio.ImageIO.read(Unknown Source)
	at javax.imageio.ImageIO.read(Unknown Source)
	at hackathon.WriteTiffTest.main(WriteTiffTest.java:16)
```````

I expect that the TIFF support was enhanced in the latest version.

I'm not sure if this is a defect or my learning curve.  Any insight would be appreciated?

Thanks,
Rich

**Comment by haraldk** [comment]
Hi Rich,

Thanks for reporting and creating a test program!

Your code seems fine! So far I have only tested the latest development version, and, according to my tests the output TIFFs are okay. They look fine, and open without problems in OS Preview. What software do you use, that can't open them?

I need some more time to see if the output is different if using the 3.2.1 release, but I know there are some minor issues fixed since then, could very well be that.

And, finally, yes. Support for CCITT encodings (TIFF compression 3 & 4) were added in 3.2, so using 3.0.x or 3.1.x won't work.


Best regards,

-- 
Harald K

**Comment by evillrich** [comment]
Hi Frank - thanks for the quick response!

Both Windows Photo Viewer and Paint, on both Windows 7 and Windows 8, just give a message that indicates that the produced TIFF file is damaged or corrupted.  Also, Paint.net, an open source image editor gives an error as well.  There was a stack trace with Paint.net, but it didn't look too meaningful in terms of details other than it is calling some MS libraries.  It's possible that all 3 of those use the same underlying libraries.

I also used GIMP and Google's Picasa image viewer and they could open the image just fine.

This is with the sample program I sent yesterday.  However, all of the image programs mentioned above have an issue if I perform a getSubimage() prior to saving.  I'll send a simple example later today that illustrates the issue.

Thanks again for your help.
Rich

**Comment by evillrich** [comment]
Hi Frank,

Here is the example I mentioned.  Let me know if you can reproduce the problem on the latest version.

Thanks,
Rich

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

import javax.imageio.ImageIO;

//Images:
//http://www.fileformat.info/format/tiff/sample/CCITT_1.TIF
//http://www.fileformat.info/format/tiff/sample/G31DS.TIF
//
public class ImageCropTest {

	public static void main(String args[]) throws IOException {

		// Read the TIFF image
		BufferedImage image = ImageIO.read(new File(args[0]));

		// Crop
		BufferedImage zone = image.getSubimage(400, 200, 800, 400);

		// Write the image out as a TIFF - writes an image that can't
		// be loaded on Windows by "Paint", "Photo viewer", etc.
		if (!ImageIO.write(zone, "TIFF", new File("zone.tif"))) {
			System.out.println("Could not write out TIFF image");
		}

		// Write the image out as a JPEG - works fine
		if (!ImageIO.write(zone, "JPEG", new File("zone.jpg"))) {
			System.out.println("Could not write out JPEG image");
		}

		// Read in the TIFF zone image we wrote out.
		BufferedImage zoneTiff = ImageIO.read(new File("zone.tif"));

		// Now write that TIFF as a JPEG - image is corrupted
		if (!ImageIO.write(zoneTiff, "JPEG", new File("zone2.jpg"))) {
			System.out.println("Could not write out 2nd JPEG image");
		}
	}
}

**Comment by haraldk** [comment]
Hi again,

First of all, I'm Harald. Not sure why you keep calling me Frank..? ;-)

Second, yes, I think I can reproduce the issue. Using the latest version, the image is written, and can be opened in my tools. However, there seems to be some issues with the cropping offsets, so the image looks all skewed. 

If you need a quick workaround, you can create a copy of the image before writing it, like this. 

````java
    public static BufferedImage createCopy(final BufferedImage pImage) {
        if (pImage == null) {
            throw new IllegalArgumentException("image == null");
        }

        ColorModel cm = pImage.getColorModel();

        BufferedImage img = new BufferedImage(cm,
                                              cm.createCompatibleWritableRaster(pImage.getWidth(), pImage.getHeight()),
                                              cm.isAlphaPremultiplied(), null);

        Graphics2D g = img.createGraphics();
        try {
            g.setComposite(AlphaComposite.Src);
            g.setRenderingHint(RenderingHints.KEY_DITHERING, RenderingHints.VALUE_DITHER_DISABLE);
            g.drawImage(pImage, 0, 0, null);
        }
        finally {
            g.dispose();
        }

        return img;
    }
````


Regards,

-- 
Harald K

**Comment by evillrich** [comment]
Big apologies Harald.  Working too many late nights I guess.  ;)  Not sure where I got Frank from.

I'll try the workaround.  Really appreciate the great support you provide.

Thanks

**Comment by evillrich** [comment]
Latest update.  The workaround did allow me to save the cropped results as JPEG.  Thanks for supplying.

For my use case, I really need to save as TIFF.  And Windows tools don't like the TIFF output of TwelveMonkeys.  Even if I look at the properties of the TIFF file in Windows, it shows none of the TIFF attributes... 

I ran the TIFF image through the tiffinfo tool that is part of libtiff and it gave me this info:
TIFFReadDirectory: Warning, test.tif: **Bogus "StripByteCounts" field, ignoring and calculating from imagelength.**
TIFF Directory at offset 0x40 (64)
  Image Width: 2464 Image Length: 3248
  Resolution: 72, 72 pixels/inch
  Bits/Sample: 1
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 2147483647
  Planar Configuration: single image plane
  Software: TwelveMonkeys ImageIO TIFF writer 3.2.1
 
I wonder if that warning is an issue.  Also, I used tiffcp to concatenate 2 of the images output from TwelveMonkeys and it produced a TIFF that can be viewed on Windows.  So, libtiff is writing different output that can be digested by Windows.

Hope that info helps.
Rich

**Comment by haraldk** [comment]
Hmmm..

There seems to be an issue that the writer does not take into account that the samples can be less than 8 bit/sample... This results in the `StripByteCounts` field getting an incorrect value (as spotted by `tifftool`/LibTIFF).

My own reader, and all other software I normally use copes with this, so I haven't noticed this issue before. But of course, we want to be spec compliant. Thanks again for reporting and providing extra info! I'll have to look into this issue some more, and will hopefully have a fix available soon.


-- 
Harald K

**Comment by evillrich** [comment]
Thanks Harald.

The TIFF output issue and the cropping issue are 2 different issues then, correct?  Just want to make sure I'm following.  

Rich

**Comment by haraldk** [comment]
Correct. 

1. When you crop the image using `BufferedImage.getSubimage(..)`, the new "sub" image will share pixel data with the original. Only the raster will have an offset and different dimensions. As the writer is accessing the pixel data directly without taking the offset and the original pixel stride into account, the written image looked skewed (or in other words, just completely messed up ;-) ). The `createCopy(..)` method above is a workaround for this issue.

2. In addition, the writer didn't correctly calculate `StripByteCounts` for uncompressed TIFFs (the default) if the bits per sample was not 8. You might be able to work around this, by using LZW or PackBits compression. However, most tools will just ignore the value (as in the "Warning: Bogus StripByteCounts field"... above) and compute it correctly anyway if the data is uncompressed. 

I've created test cases for the issues and fixed some locally. You'll find an update here as soon as I have pushed a fix.

-- 
Harald K
